src/builtin.jq

```diff
@@ -261,6 +269,17 @@ def walk(f):
   else f
   end;
 
+# Run `protect` no matter what when backtracking through a call to this
+# function.  `protect` gets called with `true` if no error was raised,
+# else with the error (wrapped in an array) that was raised.
+def _unwind(protect):
+  . as $dot |
+  unwinding |
+  if .==false then $dot
+  else protect
+  end;
+def unwind(protect): _unwind(protect|empty);
```

Comment:
I have a hard time understanding the naming choice for the method and for the instruction, too. Why unwind?
There is an instruction (and a whole concept) `BACKTRACK` but the corresponding builtin is called `empty`. I believe this is a great example of user-facing syntax hiding internal implementation details. 
I'd love to see something similar here, but can't yet think of anything good

Reply:
Lisp.  Lisp has `unwind-protect`.  This is `def unwind(protect): ...`.  The idea is to run the `protect` closure on unwind.  Granted, in Lisp the "protect" is part of the name of the primitive.  Maybe I should call it `unwind_protect`?

Reply:
OK. I'm not familiar with LISP, but maybe simple `protect` will be meaningful to everyone?

Reply:
Maybe `on_unwind/1`?

Reply:
`on_unwind` sounds good.